src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+	(												\
+		(((((UINT64) (x)) >> 56) & ((UINT64) 0xFF)) <<  0)	|	\
+		(((((UINT64) (x)) >> 48) & ((UINT64) 0xFF)) <<  8)	|	\
+		(((((UINT64) (x)) >> 40) & ((UINT64) 0xFF)) << 16)	|	\
+		(((((UINT64) (x)) >> 32) & ((UINT64) 0xFF)) << 24)	|	\
+		(((((UINT64) (x)) >> 24) & ((UINT64) 0xFF)) << 32)	|	\
+		(((((UINT64) (x)) >> 16) & ((UINT64) 0xFF)) << 40)	|	\
+		(((((UINT64) (x)) >>  8) & ((UINT64) 0xFF)) << 48)	|	\
+		(((((UINT64) (x)) >>  0) & ((UINT64) 0xFF)) << 56)		\
+	)
+
+#define ACCESSING_BITS_0_15				((mem_mask & 0x0000ffff) != 0)
+#define ACCESSING_BITS_16_31			((mem_mask & 0xffff0000) != 0)
+
+//// constants for expression endianness
+#ifndef WORDS_BIGENDIAN
+#define NATIVE_ENDIAN_VALUE_LE_BE(leval,beval)	(leval)
+#else
+#define NATIVE_ENDIAN_VALUE_LE_BE(leval,beval)	(beval)
+#endif
+
+#define BYTE4_XOR_LE(a) 				((a) ^ NATIVE_ENDIAN_VALUE_LE_BE(0,3))
+
+#define BYTE_XOR_LE(a)  				((a) ^ NATIVE_ENDIAN_VALUE_LE_BE(0,1))
+
+#define profiler_mark_start(x)	do { } while (0)
+#define profiler_mark_end()		do { } while (0)
+
+/* Highly useful macro for compile-time knowledge of an array size */
+#define ARRAY_LENGTH(x)		(sizeof(x) / sizeof(x[0]))
+
+inline INT32 mul_32x32_shift(INT32 a, INT32 b, INT8 shift)
+{
+	return (INT32)(((INT64)a * (INT64)b) >> shift);
+}
+
+#if !defined(__SSE2__) && (_M_IX86_FP == 2 || (defined(_M_AMD64) || defined(_M_X64)))
+#define __SSE2__ 1
+#endif
+#if defined(__SSE2__) && __SSE2__
+#include <emmintrin.h>
+static INT16 sse2_scale_table[256][8];
+#endif
+
+inline rgb_t rgba_bilinear_filter(rgb_t rgb00, rgb_t rgb01, rgb_t rgb10,
+                                  rgb_t rgb11, UINT8 u, UINT8 v)
+{
+#if defined(__SSE2__) && __SSE2__
+	__m128i  scale_u = *(__m128i *)sse2_scale_table[u], scale_v = *(__m128i *)sse2_scale_table[v];
+	return _mm_cvtsi128_si32(_mm_packus_epi16(_mm_packs_epi32(_mm_srli_epi32(_mm_madd_epi16(_mm_max_epi16(
+		_mm_slli_epi32(_mm_madd_epi16(_mm_unpacklo_epi8(_mm_unpacklo_epi8(_mm_cvtsi32_si128(rgb01), _mm_cvtsi32_si128(rgb00)), _mm_setzero_si128()), scale_u), 15), 
+		_mm_srli_epi32(_mm_madd_epi16(_mm_unpacklo_epi8(_mm_unpacklo_epi8(_mm_cvtsi32_si128(rgb11), _mm_cvtsi32_si128(rgb10)), _mm_setzero_si128()), scale_u), 1)),
+		scale_v), 15), _mm_setzero_si128()), _mm_setzero_si128()));
+#else
+	UINT32 ag0, ag1, rb0, rb1;
+	rb0 = (rgb00 & 0x00ff00ff) + ((((rgb01 & 0x00ff00ff) - (rgb00 & 0x00ff00ff)) * u) >> 8);
+	rb1 = (rgb10 & 0x00ff00ff) + ((((rgb11 & 0x00ff00ff) - (rgb10 & 0x00ff00ff)) * u) >> 8);
+	rgb00 >>= 8;
+	rgb01 >>= 8;
+	rgb10 >>= 8;
+	rgb11 >>= 8;
+	ag0 = (rgb00 & 0x00ff00ff) + ((((rgb01 & 0x00ff00ff) - (rgb00 & 0x00ff00ff)) * u) >> 8);
+	ag1 = (rgb10 & 0x00ff00ff) + ((((rgb11 & 0x00ff00ff) - (rgb10 & 0x00ff00ff)) * u) >> 8);
+	rb0 = (rb0 & 0x00ff00ff) + ((((rb1 & 0x00ff00ff) - (rb0 & 0x00ff00ff)) * v) >> 8);
+	ag0 = (ag0 & 0x00ff00ff) + ((((ag1 & 0x00ff00ff) - (ag0 & 0x00ff00ff)) * v) >> 8);
+	return ((ag0 << 8) & 0xff00ff00) | (rb0 & 0x00ff00ff);
+#endif
```

Comment:
I'd measure performance without the SSE intrinsic stuff; maybe the regular C code translates to fast enough assembly, or maybe the compiler can vectorise it automatically. In any case, M1 would not benefit from that.

Reply:
Agreed, to be fixed in future PR.